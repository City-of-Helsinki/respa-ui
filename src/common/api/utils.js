import get from 'lodash/get';

import constants from '../../../app/constants/AppConstants';
import store from '../../store';

let authToken;

store.subscribe(() => {
  const state = store.getState();
  authToken = get(state, 'auth.token');
});

function getAuthorizationHeader() {
  return authToken
    ? { Authorization: `JWT ${authToken}` }
    : {};
}

/**
 * Getter for API url.
 * @param endpoint
 * @param params
 * @param trailingSlash This is because Django adds trailing slash into every endpoint.
 * @returns {string}
 */
export const getUrl = (endpoint, params = {}, trailingSlash = true) => {
  const query = Object.keys(params).map(key => `${key}=${params[key]}`).join('&');
  return `${constants.API_URL}/${endpoint}${trailingSlash ? '/' : ''}${query ? `?${query}` : ''}`;
};

/**
 * Getter for default headers.
 * @returns {object}
 */
export const getHeaders = () => ({
  ...constants.REQUIRED_API_HEADERS,
  ...getAuthorizationHeader(),
});
