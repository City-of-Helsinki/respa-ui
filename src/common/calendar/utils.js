export const getDefaultSelectedTimeRange = (reservation) => {
  return reservation ? {
    start: reservation.begin,
    end: reservation.end
  } : null;
};
