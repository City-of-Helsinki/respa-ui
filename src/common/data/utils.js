import get from 'lodash/get';
import values from 'lodash/values';
import pickBy from 'lodash/pickBy';

export const getLocalizedFieldValue = (field, locale, fallback = false) => {
  const localeValue = get(field, locale, null);

  if (localeValue || !fallback) {
    return localeValue;
  }

  return values(field).find(fallbackValue => !!fallbackValue);
};


export const parseData = (data) => {
  return pickBy(data, value => value || value === 0);
};
