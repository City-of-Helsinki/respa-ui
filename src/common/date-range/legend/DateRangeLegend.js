import React from 'react';
import PropTypes from 'prop-types';

import injectT from '../../../../app/i18n/injectT';

const DateRangeLegend = (props) => {
  const { t } = props;
  return (
    <div className="app-DateRangeLegend">
      <div className="app-DateRangeLegend__legend-item">
        <div className="app-DateRangeLegend__color-box app-DateRangeLegend__color-box--available">01</div>
        <p>{t('DateRangeLegend.available')}</p>
      </div>

      <div className="app-DateRangeLegend__legend-item">
        <div className="app-DateRangeLegend__color-box app-DateRangeLegend__color-box--start">01</div>
        <p>{t('DateRangeLegend.availableStartingDay')}</p>
      </div>

      <div className="app-DateRangeLegend__legend-item">
        <div className="app-DateRangeLegend__color-box app-DateRangeLegend__color-box--reserved">01</div>
        <p>{t('DateRangeLegend.reserved')}</p>
      </div>

      <div className="app-DateRangeLegend__legend-item">
        <div className="app-DateRangeLegend__color-box app-DateRangeLegend__color-box--selection">01</div>
        <p>{t('DateRangeLegend.selection')}</p>
      </div>

      <div className="app-DateRangeLegend__legend-item">
        <div className="app-DateRangeLegend__color-box app-DateRangeLegend__color-box--unavailable">01</div>
        <p>{t('DateRangeLegend.unavailable')}</p>
      </div>
    </div>
  );
};

DateRangeLegend.propTypes = {
  t: PropTypes.func.isRequired,
};

export default injectT(DateRangeLegend);
