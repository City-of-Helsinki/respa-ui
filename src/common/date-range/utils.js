import moment from 'moment';
import { DateUtils } from 'react-day-picker';

import AppConstants from '../../../app/constants/AppConstants';

export const DatePickerValidationError = Object.freeze({
  RANGE_TOO_SHORT: 'range_too_short',
  RANGE_TOO_LONG: 'range_too_long',
  INVALID_STARTING_DAY: 'invalid_starting_day',
  OVERLAPPING_RESERVATION: 'overlapping_reservation',
  OUTSIDE_OPENING_PERIODS: 'outside_opening_periods',
});

const diffDays = (date, other) => {
  return moment(date).startOf('day').diff(moment(other).startOf('day'), 'days') + 1;
};

const isRangeTooLong = (range, maxDays) => {
  return diffDays(range.to, range.from) > maxDays;
};

const isRangeTooShort = (range, minDays) => {
  return diffDays(range.to, range.from) < minDays;
};

const isValidStartingDay = (range, startingDays) => {
  return startingDays.some(startingDay => moment(startingDay).isSame(range.from, 'day'));
};

const isOverlappingReservation = (range, reservations) => {
  return reservations.some(
    reservation => DateUtils.isDayInRange(range.from, reservation)
      || DateUtils.isDayInRange(range.to, reservation)
      || DateUtils.isDayInRange(reservation.from, range)
      || DateUtils.isDayInRange(reservation.to, range),
  );
};

const isWithinOpeningPeriods = (range, openingPeriods) => {
  // TODO: perhaps need to check for overlapping opening periods and merge them?
  return openingPeriods.some(
    openingPeriod => DateUtils.isDayInRange(range.from, openingPeriod)
      && DateUtils.isDayInRange(range.to, openingPeriod),
  );
};

export const validateRange = (range, constraints) => {
  const validationErrors = [];
  if (isOverlappingReservation(range, constraints.reservations)) {
    validationErrors.push(DatePickerValidationError.OVERLAPPING_RESERVATION);
  }
  if (!isValidStartingDay(range, constraints.startingDays)) {
    validationErrors.push(DatePickerValidationError.INVALID_STARTING_DAY);
  }
  if (isRangeTooLong(range, constraints.maxDays)) {
    validationErrors.push(DatePickerValidationError.RANGE_TOO_LONG);
  }
  if (isRangeTooShort(range, constraints.minDays)) {
    validationErrors.push(DatePickerValidationError.RANGE_TOO_SHORT);
  }
  if (!isWithinOpeningPeriods(range, constraints.openingPeriods)) {
    validationErrors.push(DatePickerValidationError.OUTSIDE_OPENING_PERIODS);
  }
  return validationErrors;
};

export const isRange = (range) => {
  return range.from && range.to;
};

export const formatDateRange = (startDate, endDate) => {
  if (!startDate || !endDate) return '-';

  return `${moment(startDate).format(AppConstants.DATE_FORMAT)} - ${moment(
    endDate,
  ).format(AppConstants.DATE_FORMAT)}`;
};

export const formatTimeRange = (startDate, endDate) => {
  if (!startDate || !endDate) return '-';

  return `${moment(startDate).format('HH:mm')} - ${moment(endDate).format(
    'HH:mm',
  )}`;
};

export const getDateRangeDuration = (startDate, endDate) => {
  return moment(endDate).diff(moment(startDate), 'days');
};
