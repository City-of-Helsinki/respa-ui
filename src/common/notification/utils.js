import { NotificationManager } from 'react-notifications';

import { NOTIFICATION_TYPE } from './constants';

export const createNotification = (type, message = '', title = '', fadeTime = 2000, callback) => {
  switch (type) {
    case NOTIFICATION_TYPE.INFO:
      NotificationManager.info(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.SUCCESS:
      NotificationManager.success(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.WARNING:
      NotificationManager.warning(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.ERROR:
      NotificationManager.error(message, title, fadeTime, callback);
      break;
    default:
      NotificationManager.info(message, title, fadeTime, callback);
  }
};
