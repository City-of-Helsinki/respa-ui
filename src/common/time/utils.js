import moment from 'moment';

import { DEFAULT_TIME_FORMAT } from './constants';

export const momentGetter = arg => moment(arg);

export const formatTime = (time, formatString) => {
  return momentGetter(time).format(formatString || DEFAULT_TIME_FORMAT);
};
