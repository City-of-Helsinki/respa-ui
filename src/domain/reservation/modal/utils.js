export const getCancelCategories = (data = [], isAdmin, locale) => {
  return data
    .filter(({ reservation_type: reservationType }) => (
      isAdmin ? true : reservationType === 'own'))
    .map(category => ({
      value: category.id,
      label: category.name[locale || 'fi'],
    }));
};
