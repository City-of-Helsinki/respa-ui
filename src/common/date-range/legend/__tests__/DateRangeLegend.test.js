import React from 'react';
import toJson from 'enzyme-to-json';

import DateRangeLegend from '../DateRangeLegend';
import { shallowWithIntl } from '../../../../../app/utils/testUtils';

describe('DateRangeLegend', () => {
  const getWrapper = () => shallowWithIntl(<DateRangeLegend />);

  test('renders normally', () => {
    const wrapper = getWrapper();
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
