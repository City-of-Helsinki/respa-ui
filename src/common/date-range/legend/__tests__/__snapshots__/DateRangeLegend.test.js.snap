// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateRangeLegend renders normally 1`] = `
<div
  className="app-DateRangeLegend"
>
  <div
    className="app-DateRangeLegend__legend-item"
  >
    <div
      className="app-DateRangeLegend__color-box app-DateRangeLegend__color-box--available"
    >
      01
    </div>
    <p>
      DateRangeLegend.available
    </p>
  </div>
  <div
    className="app-DateRangeLegend__legend-item"
  >
    <div
      className="app-DateRangeLegend__color-box app-DateRangeLegend__color-box--start"
    >
      01
    </div>
    <p>
      DateRangeLegend.availableStartingDay
    </p>
  </div>
  <div
    className="app-DateRangeLegend__legend-item"
  >
    <div
      className="app-DateRangeLegend__color-box app-DateRangeLegend__color-box--reserved"
    >
      01
    </div>
    <p>
      DateRangeLegend.reserved
    </p>
  </div>
  <div
    className="app-DateRangeLegend__legend-item"
  >
    <div
      className="app-DateRangeLegend__color-box app-DateRangeLegend__color-box--selection"
    >
      01
    </div>
    <p>
      DateRangeLegend.selection
    </p>
  </div>
  <div
    className="app-DateRangeLegend__legend-item"
  >
    <div
      className="app-DateRangeLegend__color-box app-DateRangeLegend__color-box--unavailable"
    >
      01
    </div>
    <p>
      DateRangeLegend.unavailable
    </p>
  </div>
</div>
`;
